File: 6/python/receiver_attch/main.py
import json
import imaplib
import email
import os

def load_config(filename="../../config.json"):
    with open(filename, "r", encoding="utf-8") as f:
        return json.load(f)

def fetch_attachment(config):
    imap_server = config["imapServer"]
    username = config["username"]
    password = config["password"]

    if ":" in imap_server:
        host, port = imap_server.split(":")
        port = int(port)
    else:
        host = imap_server
        port = 993

    mail_server = imaplib.IMAP4_SSL(host, port)
    mail_server.login(username, password)
    mail_server.select("INBOX")

    typ, data = mail_server.search(None, "ALL")
    if typ != "OK":
        print("Ошибка поиска писем")
        return

    msg_nums = data[0].split()
    last_ten = msg_nums[-10:] if len(msg_nums) >= 10 else msg_nums

    attachment_found = False
    for num in last_ten:
        typ, msg_data = mail_server.fetch(num, "(RFC822)")
        if typ != "OK":
            print("Ошибка получения письма", num)
            continue
        msg = email.message_from_bytes(msg_data[0][1])
        for part in msg.walk():
            content_disposition = part.get("Content-Disposition", "")
            if "attachment" in content_disposition.lower():
                filename = part.get_filename()
                if filename:
                    with open(filename, "wb") as f:
                        f.write(part.get_payload(decode=True))
                    print(f"Вложение сохранено в файл: {filename}")
                    attachment_found = True
                    break
        if attachment_found:
            break

    if not attachment_found:
        print("Письмо с вложением не найдено среди последних писем.")
    mail_server.logout()

if __name__ == '__main__':
    config = load_config()
    fetch_attachment(config)

File: 6/receiver_attch/main.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"

	"github.com/emersion/go-imap"
	"github.com/emersion/go-imap/client"
	"github.com/emersion/go-message/mail"
)

type Config struct {
	SMTPHost      string `json:"smtpHost"`
	SMTPPort      string `json:"smtpPort"`
	IMAPServer    string `json:"imapServer"`
	Username      string `json:"username"`
	Password      string `json:"password"`
	From          string `json:"from"`
	To            string `json:"to"`
	LabName       string `json:"labName"`
	SenderName    string `json:"senderName"`
	SenderGroup   string `json:"senderGroup"`
	ReceiverName  string `json:"receiverName"`
	ReceiverGroup string `json:"receiverGroup"`
}

func loadConfig(filename string) (*Config, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	var config Config
	if err := json.Unmarshal(data, &config); err != nil {
		return nil, err
	}
	return &config, nil
}

func fetchEmailWithAttachment(config *Config) error {
	c, err := client.DialTLS(config.IMAPServer, nil)
	if err != nil {
		return err
	}
	defer c.Logout()

	if err := c.Login(config.Username, config.Password); err != nil {
		return err
	}

	mbox, err := c.Select("INBOX", false)
	if err != nil {
		return err
	}

	var seqStart uint32 = 1
	if mbox.Messages > 10 {
		seqStart = mbox.Messages - 10 + 1
	}
	seqset := new(imap.SeqSet)
	seqset.AddRange(seqStart, mbox.Messages)

	messages := make(chan *imap.Message, 10)
	section := &imap.BodySectionName{}
	go func() {
		if err := c.Fetch(seqset, []imap.FetchItem{imap.FetchEnvelope, section.FetchItem()}, messages); err != nil {
			log.Fatal(err)
		}
	}()

	for msg := range messages {
		r := msg.GetBody(section)
		if r == nil {
			continue
		}
		mr, err := mail.CreateReader(r)
		if err != nil {
			continue
		}
		for {
			p, err := mr.NextPart()
			if err == io.EOF {
				break
			} else if err != nil {
				break
			}
			switch h := p.Header.(type) {
			case *mail.AttachmentHeader:
				filename, _ := h.Filename()
				f, err := os.Create(filename)
				if err != nil {
					return err
				}
				defer f.Close()
				if _, err := io.Copy(f, p.Body); err != nil {
					return err
				}
				fmt.Printf("Вложение сохранено в файл: %s\n", filename)
				return nil
			}
		}
	}
	fmt.Println("Письмо с вложением не найдено среди последних писем.")
	return nil
}

func main() {
	config, err := loadConfig("../config.json")
	if err != nil {
		log.Fatal("Ошибка загрузки конфигурации:", err)
	}

	if err := fetchEmailWithAttachment(config); err != nil {
		log.Fatal("Ошибка при получении вложения:", err)
	}
}
